File: 967-minimum-falling-path-sum/minimum-falling-path-sum.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function (matrix) {
    const n = matrix.length
    let memo = new Array(n).fill().map(() => new Array(n).fill(null));

    let dp = (row, col) => {
        // base cases
        if (col < 0 || col >= n) return Infinity;
        if (row === n - 1) return matrix[row][col];
        if (memo[row][col] !== null) return memo[row][col];
        let down = dp(row + 1, col);
        let left = dp(row + 1, col - 1);
        let right = dp(row + 1, col + 1);

        memo[row][col] = matrix[row][col] + Math.min(down, left, right);
        return memo[row][col];
    }
    let minFallingSum = Infinity;
    for (let col = 0; col < n; col++) {
        minFallingSum = Math.min(minFallingSum, dp(0, col));
    }
    return minFallingSum;
};

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int ans = Integer.MAX_VALUE;

        if(n < 2) return matrix[0][0];

        for(int i = 1; i < n; i++){
            for(int j = 0; j < n; j++){

                int min = matrix[i - 1][j];
                if(j - 1 >= 0)
                    min = Math.min(min, matrix[i - 1][j - 1]);
                if(j + 1 < n)
                    min = Math.min(min,  matrix[i - 1][j + 1]);
                
                matrix[i][j] += min;
               
               if(i == n - 1)
                ans = Math.min(ans, matrix[i][j]);
            }
        }
        return ans;
    }
}